test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -333,29 +310,28 @@ void waitForTxnsToComplete() {
         }
     }
 
-
-    CompletableFuture<Void> startWritingIntoTxn(final EventStreamWriter<Long> writer) {
+    CompletableFuture<Void> startWritingIntoTxn(final EventStreamWriter<String> writer) {
         return CompletableFuture.runAsync(() -> {
             while (!testState.stopWriteFlag.get()) {
-                Transaction<Long> transaction = null;
+                Transaction<String> transaction = null;
                 AtomicBoolean txnIsDone = new AtomicBoolean(false);
 
                 try {
-                    List<Long> eventsInTxn = new ArrayList<>(NUM_EVENTS_PER_TRANSACTION);
                     transaction = writer.beginTxn();
-
+                    final String uniqueRoutingKey = transaction.getTxnId().toString();
                     for (int j = 0; j < NUM_EVENTS_PER_TRANSACTION; j++) {
-                        long value = testState.getNext();
-                        eventsInTxn.add(value);
-                        transaction.writeEvent(String.valueOf(value), value);
-                        log.debug("Writing event: {} into transaction: {}", value, transaction.getTxnId());
+                        // The content of events is generated following the pattern routingKey:seq_number. In this case,
+                        // the context of the routing key is the transaction.
+                        transaction.writeEvent(uniqueRoutingKey, uniqueRoutingKey + RK_VALUE_SEPARATOR + j);
```

Comment:
Suggestion: The older code would write events based on the constantly changing routing key. i.e. events are written into more than one segment , with this change a transaction contains data for a single segment alone. 
Could we use multiple keys for a given writer both in a txn and non-txn case?

Reply:
Thanks for the suggestion. Now both writers and transactions encompass multiple routing keys to write to multiple segments (this can be defined in the test).